[33617] Allow SSO providers to define logout flows

oliverguenther opened the pull request:
This PR provides the following changes:

- Session values when logging in can be retained by properties in the omniauth/SSO provider configuration. This ensures that custom session values such as `saml_uid` and `saml_session_index` are being retained for single log out purposes.
- Providers are called with a signing out callback just after `AccountController#logout` was performed. They get the previous session to ensure they can restore some of these custom values such as `saml_uid` for redirecting back to the idP.

https://community.openproject.com/wp/33617

oliverguenther committed:
Allow to find omniauth strategy by name

oliverguenther committed:
Retain session values from a provider

oliverguenther committed:
Add spec to test taking over session values

oliverguenther committed:
Add test that slo callback is being called

oliverguenther committed:
Move provider finder into auth plugin

machisuji commented on app/controllers/concerns/accounts/current_user.rb:
```diff
@@ -103,13 +103,19 @@ def logout_user
   # Redirect the user according to the logout scheme
-  def perform_post_logout(_prev_session)
+  def perform_post_logout(prev_session, prev_user)
     if Setting.login_required? && omniauth_direct_login?
```
Originally this block is only there to prevent an instant re-login after logging out with direct login enabled.
Even with it enabled the user should be redirected to the configured slo_target_url, though, shouldn't they?

If the SSO provider then redirects back to OpenProject we have to make sure that OpenProject doesn't try to initiate another login in this case.

oliverguenther replied:
I think even in the case that OpenProject decides to redirect to the idP again, if you were globally logged out, you would only be redirected to the login form so I believe this would be o kay.

In any case, an omniauth direct login should still be allowing the SLO callback. Will fix that

oliverguenther committed:
Test for direct login only after the logout callback was called

This allows single-log-out flows to happen even if you
assigned an omniauth direct login.

If you came back from the provider with no SLO happening (a redirect),
it will perform the previous check to prevent you being logged in immediately again.

machisuji commented on spec/services/users/login_service_spec.rb:
```diff
@@ -0,0 +1,79 @@
+
+        it 'retains present session values' do
+          session[:foo] = 'foo value'
+          session[:what] = 'should be cleared'
```
You are not confirming that this has been cleared below.

machisuji reviewed (commented):
Looks good overall. Just one minor remark in the login service spec. Also could you please add a test case for the direct login scenario just in case?

oliverguenther committed:
extend spec to check for removed value

oliverguenther committed:
Add spec for direct login case
